### PR TITLE
Support Decimals and Date32 in Avro format

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support Decimals and Date32 in Avro format

### CODE REVIEW

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp [hunk 1]
```diff
@@ -153,6 +157,39 @@ static void insertNumber(IColumn & column, WhichDataType type, T value)
...
+            "Cannot insert Avro decimal with scale {} and precision {} to ClickHouse Decimal with scale {} and precision {}",
+            logical_type.scale(),
+            logical_type.precision(),
+            decimal_type.getScale(),
+            decimal_type.getPrecision());
+
+    return [tmp = std::string(), target_type](IColumn & column, avro::Decoder & decoder) mutable
+    {
+        decoder.decodeString(tmp);
+        if (tmp.size() != sizeof(typename DecimalType::FieldType))
```
Comment (antonio2368) [inline src/Processors/Formats/Impl/AvroRowInputFormat.cpp:177]: is it possible to create
```
static constexpr size_t field_type_size = sizeof(typename DecimalType::FieldType)
```
🤔

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp [hunk 2]
```diff
@@ -153,6 +157,39 @@ static void insertNumber(IColumn & column, WhichDataType type, T value)
...
+    if (decimal_type.getScale() != static_cast<UInt32>(logical_type.scale()) || decimal_type.getPrecision() != static_cast<UInt32>(logical_type.precision()))
+        throw Exception(
+            ErrorCodes::BAD_ARGUMENTS,
+            "Cannot insert Avro decimal with scale {} and precision {} to ClickHouse Decimal with scale {} and precision {}",
+            logical_type.scale(),
+            logical_type.precision(),
+            decimal_type.getScale(),
+            decimal_type.getPrecision());
+
+    return [tmp = std::string(), target_type](IColumn & column, avro::Decoder & decoder) mutable
```
Comment (antonio2368) [inline src/Processors/Formats/Impl/AvroRowInputFormat.cpp:174]: why do we do this with `tmp`?
Why not create a temp variable on each call?

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp [hunk 3]
```diff
@@ -1042,19 +1087,40 @@ DataTypePtr AvroSchemaReader::avroNodeToDataType(avro::NodePtr node)
...
+                return {std::make_shared<DataTypeDate32>()};
+
             return {std::make_shared<DataTypeInt32>()};
+        }
         case avro::Type::AVRO_LONG:
+        {
+            auto logical_type = node->logicalType();
+            if (logical_type.type() == avro::LogicalType::TIMESTAMP_MILLIS)
+                return {std::make_shared<DataTypeDateTime64>(3)};
+            else if (logical_type.type() == avro::LogicalType::TIMESTAMP_MICROS)
```
Comment (antonio2368) [inline src/Processors/Formats/Impl/AvroRowInputFormat.cpp:1101]:
```suggestion
            if (logical_type.type() == avro::LogicalType::TIMESTAMP_MICROS)
```

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp (hunk 1 above)
Comment (Avogar) [reply src/Processors/Formats/Impl/AvroRowInputFormat.cpp]: Yes, done.

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp (hunk 2 above)
Comment (antonio2368) [reply src/Processors/Formats/Impl/AvroRowInputFormat.cpp:174]: yes it makes sense, I thought I saw move from it somewhere but it seems I'm crazy

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**